### DISCUSSION TITLE
Get History of Revisions of kind:Kustomization

### DISCUSSION THREAD

**Discussion by ccs1910** [fluxcd/flux2#4240] [Q&A]
Hello all, 

we have a case, 
when one of my colleagues registered a new source ( " flux create source git ") on Sept 6th, 2023 at 20:00. Apparently, all k8s objects under a particular namespace were destroyed and recreated at the same time (Sept 6th, @ 20:00).

the anomaly, we don't find any changes on the code repo, related to "SourceRef".

So, we suspect the "Source Registration" destroyed the k8s resources. 

Questions:
1. Confirm if the command "flux create source git" will not change the SourceRef param, therefore kustomization will not pull from new repo ?

2. Is there any flux cli command to pull history of "revision" on kind: Kustomization ?

thanks,
Christian

**Comment by makkes** [comment]
All `flux create source git` does is create or update a `GitRepository` object on the cluster.

The only way I can imagine how this may lead to resources being deleted is that when you pass a name for which a GitRepository already exists:

```
flux create source git test --url https://github.com/org/repo --branch=main
```

will create GitRepository "test". Running

```
flux create source git test --url https://github.com/anotherorg/anotherrepo --branch=dev
```

will update the "test" GitRepository. If you have e.g. a Kustomization in your cluster that refers to the "test" GitRepository from its `.spec.sourceRef` field this may lead to resources being deleted.

**Reply by kingdonb** [reply]
This sounds plausible to me. There are some safeguards, if Kustomization is pointed at a specific path and it finds that path is not extant in the git repository listed in `sourceRef`, then it will error out and not do anything. However if it finds an empty directory at that path, and is able to sync from the empty directory, it will garbage collect any resources since it doesn't detect any error, it perceives those resources as deleted from the repo and has no way to know that you did not want those resources to be deleted in the cluster.

You can maintain a history of resource modifications based on the monitoring described with kube-state-metrics in the new monitoring guide, you can add any field from the CRD spec or status to a metric and then you'll have an audit history of those values. This is similar to the example [custom metric labels](https://fluxcd.io/flux/monitoring/custom-metrics/#adding-custom-metric-labels), if you have values you are interested in preserving the history of fields like git branch, revision, etc. then be sure you have kube-state-metrics configured first, then create a custom metric for that field.

> the anomaly, we don't find any changes on the code repo, related to "SourceRef".

I wouldn't expect to see any changes in the source code in the repo from running `flux create` commands.

Those commands run directly in the cluster, and they modify the cluster state without going through Git.

We always recommend using `flux create --export` and saving the content to the Git repository, rather than changing the cluster state directly. You can set up your production systems so that developers do not have direct access and they can only do this through Git, we recommend that (and keep the admin kubeconfig behind a "break glass" so you can use it in an emergency when say GitHub is down, or in some other incident case.)

There may be a bug in `flux create` if running `flux create` potentially overwrites an existing object. We are talking about this in the Bug Scrub today and tested, and found it indeed does that. It seems like it should not do that, it should probably error out rather than update. The `kubectl create` command as an analogy will error out if you use it to create a resource that already exists on the cluster.

What do you think, should we change this for safety?
